### PR TITLE
Finish port to cobra

### LINKED ISSUE #512: Refactor flag parsing to support spaces

### LINKED ISSUE #759: Move to pflag/viper or cobra
### Describe the feature you'd like to request

Because we manually parse CLI args, users are getting stuck with the inflexible syntax. 

See #758 
### Describe the solution you'd like

Switch to viper and pflags for config and flags parsing 

Or

Switch to Cobra and drop mitchelh/cli. This would also let us auto generate documentation

### Describe alternatives you've considered

N/A

### PR DESCRIPTION
**Pull request by gsoltis**
Ports root command execution to `cobra`.

Introduces a couple of abstractions:
`cmdutil.Helper` -> Holds common arguments to all commands
`cmdutil.CmdBase` -> Holds subcomponents common to all commands, possibly driven by the values in `cmdutil.Helper`. Examples include `UI`, `Logger`, and the API client

Port all of our `turbo` commands to be subcommands of a root command.

Reviewers: individual commits won't build, but group related functionality. 

Fixes #759 
Fixes #512

### CODE REVIEW

**Review by chris-olszewski — APPROVED** [review]
LGTM, a lot less plumbing

File: cli/internal/cmdutil/cmdutil.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+func (h *Helper) getUI(flags *pflag.FlagSet) cli.Ui {
+	colorMode := ui.GetColorModeFromEnv()
+	if flags.Changed("no-color") && h.noColor {
+		colorMode = ui.ColorModeSuppressed
+	}
+	if flags.Changed("color") && h.forceColor {
+		colorMode = ui.ColorModeForced
+	}
+	return ui.BuildColoredUi(colorMode)
+}
```
Comment (nathanhammond) [inline cli/internal/cmdutil/cmdutil.go:87]: I just discovered that this library also implements a full command set up, not just some niceties around string formatting. We also semi-erratically use our wrapper around it in `cli/internal/ui/ui.go`.

We could stand to invest more time into this area, eventually.

File: cli/internal/cmdutil/cmdutil.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+// turbo command.
+func NewHelper(turboVersion string) *Helper {
+	return &Helper{
+		TurboVersion:   turboVersion,
+		UserConfigPath: config.DefaultUserConfigPath(),
+	}
+}
+
+// GetCmdBase returns a CmdBase instance configured with values from this helper.
+// It additionally returns a mechanism to set an error, so
```
Comment (nathanhammond) [inline cli/internal/cmdutil/cmdutil.go:148]: You accidentally a sentence.

File: cli/internal/cmdutil/cmdutil.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+		return nil, err
+	}
+	userConfig, err := config.ReadUserConfigFile(h.UserConfigPath, flags)
+	if err != nil {
+		return nil, err
+	}
+	remoteConfig := repoConfig.GetRemoteConfig(userConfig.Token())
+	if remoteConfig.Token == "" && isCI() {
+		vercelArtifactsToken := os.Getenv("VERCEL_ARTIFACTS_TOKEN")
+		vercelArtifactsOwner := os.Getenv("VERCEL_ARTIFACTS_OWNER")
```
Comment (nathanhammond) [inline cli/internal/cmdutil/cmdutil.go:175]: @tknickman I'd love to think about how to make this configurable for any platform (e.g. if they implement their own remote cache).

File: cli/internal/cmd/root.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	helper.AddFlags(flags)
+	execOpts.addFlags(flags)
+	cmd.AddCommand(login.NewLinkCommand(helper))
+	cmd.AddCommand(login.NewLoginCommand(helper))
+	cmd.AddCommand(auth.LogoutCmd(helper))
+	cmd.AddCommand(auth.UnlinkCmd(helper))
+	cmd.AddCommand(info.BinCmd(helper))
+	cmd.AddCommand(daemon.GetCmd(helper, signalWatcher))
+	cmd.AddCommand(prune.GetCmd(helper))
+	cmd.AddCommand(run.GetCmd(helper, signalWatcher))
```
Comment (nathanhammond) [inline cli/internal/cmd/root.go:121]: We should align this at some point.

**Review by nathanhammond — COMMENTED** [review]
Mostly comments and questions!

File: cli/internal/run/run.go [hunk 5]
```diff
@@ -95,29 +88,38 @@ occurred again).
 Arguments passed after '--' will be passed through to the named tasks.
 `
 
-func getCmd(config *config.Config, ui cli.Ui, signalWatcher *signals.Watcher) *cobra.Command {
+// GetCmd returns the run command
+func GetCmd(helper *cmdutil.Helper, signalWatcher *signals.Watcher) *cobra.Command {
```
[the remark this answers is not in the document]
Comment (nathanhammond) [reply cli/internal/run/run.go:92]: (This is because `signalWatcher` needs to disappear, I remembered that part of the conversation from a while ago now.)

**Pull request merged**